lattices/src/ght_lattice.rs

```diff
@@ -0,0 +1,448 @@
+        self.elements.extend(other.elements);
+        self.elements.len() > old_len
+    }
+}
+
+impl<Head, Node> PartialEq<GhtInner<Head, Node>> for GhtInner<Head, Node>
+where
+    Head: Hash + Eq + 'static + Clone,
+    Node: GeneralizedHashTrieNode + 'static + PartialEq,
+    Node::Storage: VariadicSet<Schema = Node::Schema>, // multiset is not a lattice!
+    Node::Schema: SplitBySuffix<var_type!(Head, ...Node::SuffixSchema)>,
+    GhtInner<Head, Node>: GhtGet,
+    <GhtInner<Head, Node> as GhtGet>::Get: PartialEq,
+{
+    fn eq(&self, other: &GhtInner<Head, Node>) -> bool {
+        if self.children.len() != other.children.len() {
+            return false;
+        }
+
+        for head in self.iter() {
+            let other_node = other.get(&head);
+            if other_node.is_none() {
+                return false;
+            }
+            let this_node = self.get(&head);
+            if this_node.is_none() {
+                return false;
+            }
+            if this_node.unwrap() != other_node.unwrap() {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl<Head, Node> PartialOrd<GhtInner<Head, Node>> for GhtInner<Head, Node>
+where
+    Head: Hash + Eq + 'static + Clone,
+    Node: 'static + GeneralizedHashTrieNode + PartialEq + PartialOrd,
+    Node::Storage: VariadicSet<Schema = Node::Schema>, // multiset is not a lattice!
+    Node::Schema: SplitBySuffix<var_type!(Head, ...Node::SuffixSchema)>,
+{
+    fn partial_cmp(&self, other: &GhtInner<Head, Node>) -> Option<Ordering> {
+        if self.children.is_empty() && other.children.is_empty() {
+            Some(Equal)
+        } else {
+            // across all keys, determine if we have stuff in self that's not in other
+            // and vice versa
+            let (selfonly, mut otheronly) =
+                self.children
+                    .iter()
+                    .fold((false, false), |mut accum, (k, v)| {
+                        if !other.children.contains_key(k) {
+                            accum.0 = true; // selfonly is true
+                            accum
+                        } else {
+                            match v.partial_cmp(other.children.get(k).unwrap()) {
+                                Some(Greater) | None => {
+                                    accum.0 = true; // selfonly is true
+                                    accum
+                                }
+                                Some(Less) => {
+                                    accum.1 = true; // otheronly is true
+                                    accum
+                                }
+                                Some(Equal) => accum, // no changes
+                            }
+                        }
+                    });
+            // now check if other has keys that are missing in self
+            otheronly |= !other.children.keys().all(|k| self.children.contains_key(k));
```

Comment:
If you use a `for` loop instead of fold, then you can return early
https://github.com/hydro-project/hydroflow/blob/47cb703e771f7d1c451ceb9d185ada96410949da/lattices/src/map_union.rs#L137-L161